telemetry/internal/appinsights/inmemorychannel.go

```diff
@@ -0,0 +1,453 @@
+		}(state.buffer, state.retry, state.retryTimeout)
+	} else if state.callback != nil {
+		state.channel.signalWhenDone(state.callback)
+	}
+
+	return true
+}
+
+// Part of channel accept loop: Wait for throttle to expire while dropping messages
+func (state *inMemoryChannelState) waitThrottle() bool {
+	// Channel is currently throttled.  Once the buffer fills, messages will
+	// be lost...  If we're exiting, then we'll just try to submit anyway.  That
+	// request may be throttled and transmitRetry will perform the backoff correctly.
+
+	throttleDone := state.channel.throttle.notifyWhenReady()
+	dropped := 0
+
+	defer func() {
+		if dropped > 0 {
+			state.channel.logf("Channel dropped %d events while throttled", dropped)
+		}
+	}()
+
+	for {
+		select {
+		case <-throttleDone:
+			close(throttleDone)
+			return true
+
+		case event := <-state.channel.collectChan:
+			// If there's still room in the buffer, then go ahead and add it.
+			if len(state.buffer) < state.channel.batchSize {
+				state.buffer = append(state.buffer, event)
+			} else {
+				dropped++
+			}
+
+		case ctl := <-state.channel.controlChan:
+			if ctl.stop {
+				state.stopping = true
+				state.retry = ctl.retry
+				if !ctl.flush {
+					state.channel.signalWhenDone(ctl.callback)
+					return false
+				} else {
+					// Make an exception when stopping
+					return true
+				}
+			}
+
+			// Cannot flush
+			// TODO: Figure out what to do about callback?
+			if ctl.flush {
+				state.channel.signalWhenDone(ctl.callback)
+			}
+		}
+	}
+}
+
+// Part of channel accept loop: Clean up and close telemetry channel
+func (state *inMemoryChannelState) stop() {
+	state.channel.closed.Store(true)
+	close(state.channel.collectChan)
+	close(state.channel.controlChan)
+
+	// Throttle can't close until transmitters are done using it.
+	state.channel.waitgroup.Wait()
+	state.channel.throttle.stop()
+}
+
+func (channel *inMemoryChannel) transmitRetry(items []*contracts.Envelope, retry bool, retryTimeout time.Duration) {
+	payload, err := serialize(items)
+	if err != nil {
+		channel.log(err)
+		if payload == nil {
+			return
+		}
+	}
+	retryTimeRemaining := retryTimeout
+
+	for _, wait := range submit_retries {
+		result, err := channel.transmitter.Transmit(payload, items)
+		if err != nil {
+			channel.logf("Error transmitting payload: %v", err)
+			return
+		}
+		if result.isSuccess() {
+			return
+		}
+		if !retry {
+			return
+		}
+
+		// Check for success, determine if we need to retry anything
+		if result.canRetry() {
+			// Filter down to failed items
+			payload, items = result.getRetryItems(payload, items)
+			if len(payload) == 0 || len(items) == 0 {
+				return
+			}
+		} else {
+			channel.logf("Cannot retry telemetry submission")
+			return
+		}
+
+		// Check for throttling
+		if result.isThrottled() {
+			if !result.retryAfter.IsZero() {
+				channel.throttle.retryAfter(result.retryAfter)
+			} else {
+				// TODO: Pick a time
+			}
```

Comment:
TODO spotted. I guess currently this means that a timeless throttle doesn't delay the retry at all?

Reply:
Fixed. The retry will use the `Client.MaxBatchInterval` as delay.